### PR TITLE
Store: Hook up product category saving

### PR DESCRIPTION
**Pull request by justinshreve**
This PR adds some Redux code for updating a product category (creation code already existed). It also hooks up both actions to the UI, enabling saving for both new categories and existing categories.

Parent PR: #21351.
See: #20295.

To Test:
* Run `npm run test-client client/extensions/woocommerce/state` and make sure all tests pass.
* Go to `http://calypso.localhost:3000/store/products/category/:site` and test creating a new category.
* Select an existing category from the list `http://calypso.localhost:3000/store/products/categories/:site` and make an update to it.

### CODE REVIEW

File: client/extensions/woocommerce/app/product-categories/create.js [hunk 1]
```diff
@@ -69,17 +77,42 @@ class ProductCategoryCreate extends React.Component {
...
 	}
 
-	onSave = () => {};
+	onSave = () => {
+		const { site, category, translate } = this.props;
+		this.setState( () => ( { busy: true } ) );
+
+		const successAction = () => {
+			page.redirect( getLink( '/store/products/categories/:site', site ) );
+			return successNotice( translate( 'Category successfully created.' ), {
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/product-categories/create.js]: Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'Your post was successfully created.' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/store-product-cat-save)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by kellychoffman** [comment]
How is the order of the Categories on the list page determined? I ask because when you add a new category, hit save, you are brought back to the list page, but if you have a ton of categories you have to hunt around for the new category to double check it was indeed added. 

Would it make sense to sort it by most recently added? This might create an issue with adding a sub-level category tho. What if we also highlighted it for a couple seconds with the yellow background that we are using for new/pending reviews?

**Comment by allendav** [comment]
I get #21411 when i attempt to run the unit tests

**Comment by allendav** [comment]
Maybe need to disable the Save button or have an error if "Top level" is deselected and no parent category is selected

**Comment by allendav** [comment]
Delete-ing a category isn't hooked up yet, correct? (Clicking on delete has no effect)

**Comment by justinshreve** [comment]
@allendav Delete is a different PR. It's up for review.

File: client/extensions/woocommerce/state/data-layer/product-categories/test/index.js [hunk 2]
```diff
@@ -268,4 +272,99 @@ describe( 'handlers', () => {
 			);
 		} );
 	} );
+
+	describe( '#handleProductCategoryUpdate', () => {
+		test( 'should dispatch a post action', () => {
```
Comment (allendav) [inline client/extensions/woocommerce/state/data-layer/product-categories/test/index.js]: put

**Comment by justinshreve** [comment]
> How is the order of the Categories on the list page determined? I ask because when you add a new category, hit save, you are brought back to the list page, but if you have a ton of categories you have to hunt around for the new category to double check it was indeed added.

> Would it make sense to sort it by most recently added? This might create an issue with adding a sub-level category tho. What if we also highlighted it for a couple seconds with the yellow background that we are using for new/pending reviews?

Currently sorted by name.

Per slack, I have opened https://github.com/Automattic/wp-calypso/issues/21427.

File: client/extensions/woocommerce/state/data-layer/product-categories/test/index.js [hunk 3]
```diff
@@ -268,4 +272,99 @@ describe( 'handlers', () => {
...
+			expect( store.dispatch ).to.have.been.calledWith(
+				match( {
+					type: '%%success%%',
+					sentData: category,
+					receivedData: 'RECEIVED_DATA',
+				} )
+			);
+		} );
+
+		test( 'should dispatch a success function with extra parameters', () => {
```
Comment (allendav) [inline client/extensions/woocommerce/state/data-layer/product-categories/test/index.js]: function -> action

File: client/extensions/woocommerce/state/ui/product-categories/test/edits-reducer.js [hunk 4]
```diff
@@ -229,6 +230,30 @@ describe( 'edits-reducer', () => {
...
+		const edits1 = {
+			updates: [ category ],
+		};
+
+		const originatingAction = updateProductCategory( siteId, category );
+		const action = productCategoryUpdated( siteId, updatedCategory, originatingAction );
+
+		const edits2 = reducer( edits1, action );
+
+		expect( edits1.updates[ 0 ] ).to.eql( category );
```
[the remark this answers is not in the document]
Comment (justinshreve) [reply client/extensions/woocommerce/state/ui/product-categories/test/edits-reducer.js]: Yeah. We do this in a few other places as well I believe. Checks before/after.

**Comment by allendav** [comment]
Maybe a separate issue/pr - if you type a name for a category that already exists, maybe the error message could say that specifically

screen shot 2018-01-10 at 3 43 44 pm

File: client/extensions/woocommerce/app/product-categories/update.js [hunk 5]
```diff
@@ -76,20 +88,73 @@ class ProductCategoryUpdate extends React.Component {
...
+		accept( areYouSure, function( accepted ) {
+			if ( ! accepted ) {
+				return;
+			}
+			const successAction = () => {
+				debounce( () => {
+					page.redirect( getLink( '/store/products/categories/:site/', site ) );
+				}, 2000 )();
+				return successNotice(
+					translate( '%(name)s successfully deleted.', {
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/product-categories/update.js]: Hi! I've found a possible matching string that has already been translated **76** times:
`translate( '%s was successfully deleted.', { context: 'theme'} )` *ES Score: 7*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/store-product-cat-save)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by kellychoffman** [comment]
> Maybe a separate issue/pr - if you type a name for a category that already exists, maybe the error message could say that specifically

Great idea.

**Pull request merged**